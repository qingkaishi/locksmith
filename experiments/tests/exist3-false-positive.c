#include <stdlib.h>
#include <pthread.h>

// the lock label
#pragma existential(t,"&t.data")

pthread_mutex_t g;

struct t {

  int data;  // guarded by lock
  int gdata; // guarded by g
};

struct t a;

void doit(struct t* in) {
  start_unpack(in);
  //pthread_mutex_lock(&(in->lock));
  //in->data++;
  //pthread_mutex_unlock(&(in->lock));
  in->gdata++;
  end_unpack(in);
}

void* f(struct t* in) {
  pthread_mutex_lock(&g);
  doit(in);
  pthread_mutex_unlock(&g);
  return 0;
}

int main() {
  struct t __attribute__((packed)) *s;
  struct t *p;
  pthread_t t1, t2;

  pthread_mutex_init(&g, NULL);
  while(p) {
    s = (struct t __attribute((packed)) *) malloc(sizeof (struct t));
    //pthread_mutex_init(&s->lock, NULL);
    p = pack(s);
  }

  pthread_create(&t1, NULL, f, p);
  pthread_create(&t2, NULL, f, p);
  return 1;
}
