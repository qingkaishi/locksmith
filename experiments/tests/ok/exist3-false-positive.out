************************* STARTING *************************

shared: alloc.gdata:exist3-false-positive.c:40

Warning: Possible data race: alloc.gdata:exist3-false-positive.c:40 is not protected!
 references:
  dereference of *in.gdata:exist3-false-positive.c:17 at exist3-false-positive.c:22
    alloc.gdata:exist3-false-positive.c:40 => s->gdata:exist3-false-positive.c:34
                                           => s.gdata:exist3-false-positive.c:35 at pack exist3-false-positive.c:42
                                           => in->gdata:exist3-false-positive.c:26 at f exist3-false-positive.c:45
                                           => *in.gdata:exist3-false-positive.c:17 at doit exist3-false-positive.c:28
  locks acquired:
    <empty>
  in: main at exist3-false-positive.c:33 


*************************** DONE ***************************
